In [2]:
print("Hello, World!")


Hello, World!


Lecture 3




# Summary Notes: Python Modules and Pip

This session covers the basics of **Modules**, the two main types of modules in Python, and how to use **pip** as a package manager.

---

## 1. What are Modules?

Modules are pre-written code files that you can "borrow" and use in your own programs. Instead of writing complex logic from scratch (like advanced mathematical formulas or machine learning algorithms), you can use modules written by experts.

**Key Benefits:**

* **Saves Time:** You don't have to reinvent the wheel.
* **Reduces Errors:** Code in modules is well-tested and used by millions.
* **Focus on Big Projects:** Allows you to focus on the unique parts of your project rather than basic utility code.

---

## 2. Types of Modules

Python modules are categorized into two types:

| Type | Description | Analogy |
| --- | --- | --- |
| **Built-in Modules** | Pre-installed with Python. You don't need to download anything extra. | Household items like plates or spoons that are already in your kitchen. |
| **External Modules** | Not included by default. These are written by other people and must be installed via `pip`. | Groceries like milk or bread that you have to go out and buy when needed. |

**Examples:**

* **Built-in:** `hashlib`, `os`, `random`.
* **External:** `pandas` (Data analysis), `scikit-learn` (Machine Learning), `tensorflow`.

---

## 3. What is Pip?

**Pip** is the package manager for Python. It acts as a "delivery service" that fetches external modules from the internet and installs them on your system.

* **Windows Command:** `pip install <module_name>`
* **Mac/Linux Command:** `pip3 install <module_name>`

> **Example:** To install the Pandas library, you would type `pip install pandas` in your terminal.

---

## 4. The REPL

**REPL** stands for **Read, Evaluate, Print, Loop**.
It is an interactive shell where you can type Python code and see the results immediately. You access it by simply typing `python` (Windows) or `python3` (Mac/Linux) in your terminal.

While useful for quick testing, most development is done by writing scripts in `.py` files.

---

## 5. Coding Environments (Replit)

The video uses **Replit**, an online IDE.

* **Automation:** If you type `import pandas` in Replit, it detects the module is missing and installs it for you automatically.
* **Cross-Platform:** Since it runs in a browser, you don't have to worry about local setup or Python versions.

---

## Key Takeaways

* Use the `import` keyword to bring a module into your code.
* Always try to use the latest version of Python (currently 3.x), as the basics remain consistent.
* Python is one of the most versatile languages, used by giants like YouTube, Instagram, and Facebook.

Would you like me to create a short practice quiz based on these notes to help you test your understanding?

In [3]:
print ("hi")

hi


Lecture 4 

In [4]:
print("hello World",7)
print(5)
print("bye")

hello World 7
5
bye


## Summary Notes: Your First Python Program

### 1. The `print()` Function

* **Definition:** A built-in **function** used to display information on the screen.
* **Syntax:** You invoke a function using **parentheses** `()`.
* **Strings:** Text must be enclosed in **double quotes** `(" ")` or single quotes. Without them, Python doesn't recognize the text as a valid object (causing a *Syntax Error*).
* **Multiple Values:** You can print multiple items by separating them with a **comma** `,`.
* Example: `print("Hello", 7)` outputs `Hello 7`.



### 2. Python Scripts vs. REPL

* **REPL (Read-Evaluate-Print Loop):** An interactive environment where you type code line-by-line and get immediate results. Best for quick tests.
* **Python Script (`.py` file):** A file containing a sequence of instructions.
* **Execution Flow:** Python executes scripts **line-by-line** (top to bottom). This allows you to automate complex tasks by writing a "script" for the computer to follow.

### 3. Basic Calculations

Python can be used as a calculator within the `print()` function:

* **Addition:** `+`
* **Subtraction:** `-`
* **Multiplication:** `*` (Use the asterisk, not 'x')
* **Division:** `/`

---

### Key Takeaway

Programming is about telling the computer to perform tasks in a specific order. To succeed, you must avoid "shortcuts" and focus on **coth in a single print statement?**

Lecture 5

In [5]:
# i am shashi , and how are you 

In [6]:
print ("i am shashi
       also a good boy")

SyntaxError: unterminated string literal (detected at line 1) (2972423133.py, line 1)

In [9]:
print ("i am shashi \n also a good boy")

i am shashi 
 also a good boy


In [11]:
print ("i am shashi \nalso a good boy")

i am shashi 
also a good boy


In [12]:
'''

i am shashi 
also a good boy


'''

'\n\ni am shashi \nalso a good boy\n\n\n'

In [13]:
'''
i am shashi 
also a good boy
'''

'\ni am shashi \nalso a good boy\n'

In [15]:
'''i am shashi also a good boy'''

'i am shashi also a good boy'

In [16]:
"""hi this is shashi"""

'hi this is shashi'

In [17]:
"""
this is shashi
"""

'\nthis is shashi\n'

In [18]:
print ("i am shashi"\n and this is python")

SyntaxError: unexpected character after line continuation character (1476331931.py, line 1)

In [21]:
print ("i am \"shashi\"\nand this is python")

i am "shashi"
and this is python


In [23]:
print ("hi", 6, 7)

hi 6 7


In [22]:
print ("hi", 6, 7, sep="~")

hi~6~7


In [24]:
print ("hi", 6, 7, sep="~",end="009\n")

hi~6~7009


# Summary Notes: Comments, Escape Sequences & Print Parameters

This lesson explores how to write non-executing code, handle special characters within strings, and master the advanced features of the `print()` function.

---

## 1. Python Comments

Comments are lines of text that the Python interpreter **ignores**. They are used to explain code logic to yourself or other developers.

### Types of Comments

* **Single-line Comment:** Created using the pound/hash (`#`) symbol.
* **Multi-line Comment:** Created by enclosing text in triple quotes (`'''` or `"""`).
> *Note: While technically these are "string literals" ignored by Python when not assigned to a variable, they are commonly used as multi-line comments.*



### Helpful Shortcuts

* **Comment/Uncomment Block:** Select the lines and press `Ctrl + /` (Windows/Linux) or `Command + /` (Mac).

---

## 2. Escape Sequence Characters

To insert characters that are "illegal" or have special meaning in a string, we use a **backslash (`\`)** followed by the character.

| Escape Sequence | Description | Result |
| --- | --- | --- |
| `\n` | **New Line** | Moves the text following it to a new line. |
| `\"` | **Double Quote** | Inserts a `"` inside a string already wrapped in double quotes. |
| `\'` | **Single Quote** | Inserts a `'` inside a string already wrapped in single quotes. |

---

## 3. Advanced Print Statement

The `print()` function is more powerful than just showing text; it can take several **optional parameters**:

### A. The `sep` (Separator) Parameter

By default, when you print multiple objects separated by a comma, Python adds a space between them. Use `sep` to change this.

* **Example:** `print("Hello", "World", sep="-")` → `Hello-World`

### B. The `end` Parameter

By default, `print()` adds a new line (`\n`) at the end of its output. Use `end` to specify what happens at the very end of the statement.

* **Example:** `print("Hello", end="009")` will print "Hello009" and keep the cursor on the same line for the next print statement.

---

### Key Takeaways

* **Comments** make your code readable and maintainable over time.
* **Escape sequences** allow you to format text output (like new lines) precisely.
* **`sep` and `end**` provide control over how data  specific separator and a new line escape sequence together?

Lecture 6

In [25]:
a =1
print(a)

1


In [29]:
c = "shashi"
print(c)

shashi


In [30]:
b =a +c

TypeError: unsupported operand type(s) for +: 'int' and 'str'

In [31]:
print ("shashi"+4)

TypeError: can only concatenate str (not "int") to str

In [32]:
print (20+34)

54


In [33]:
print("type of c", type(c))

type of c <class 'str'>


In [34]:
print("type of c", type(a))

type of c <class 'int'>


In [35]:
d =1.2548
print(d)
print("type of d", type(d))

1.2548
type of d <class 'float'>


In [36]:
e =complex(4,5)
print(e)
print("type of e", type(e))

(4+5j)
type of e <class 'complex'>


In [45]:
f =None
g= TRUE
h= false

print("type of f", type(f))
print(f)
print("type of g", type(g))
print(g)
print("type of h", type(h))
print(h)

NameError: name 'TRUE' is not defined

In [44]:
f =None
g= True
h= False

print("type of f", type(f))
print(f)
print("type of g", type(g))
print(g)
print("type of h", type(h))
print(h)

type of f <class 'NoneType'>
None
type of g <class 'bool'>
True
type of h <class 'bool'>
False


In [48]:
list1=[2,3,4,5,[-3,4,-6],["shashi","kumar"]]
print(list1)


[2, 3, 4, 5, [-3, 4, -6], ['shashi', 'kumar']]


In [49]:
tuple1=(("shashi","kumar"),("xioni","LG"),(2,3,4,5))
print(tuple1)

(('shashi', 'kumar'), ('xioni', 'LG'), (2, 3, 4, 5))


In [54]:
dic1={"shashi":"kumar","xioni":"LG","age":25,"canvte":True}
print(dic1)

{'shashi': 'kumar', 'xioni': 'LG', 'age': 25, 'canvte': True}


# Summary Notes: Variables and Data Types

Welcome to Day 6! Today’s session focuses on how Python stores information in memory and the different "flavors" of data we can use.

---

## 1. What is a Variable?

A **Variable** is a container for storing data values.

* **Analogy:** Think of a kitchen container. It can hold sugar today, but you can empty it and put salt in it tomorrow.
* **How it works:** When you create a variable (e.g., `a = 1`), Python allocates a space in your **RAM** (memory) to store that value and labels it with the name `a`.

---

## 2. Data Types

A **Data Type** specifies the category of data a variable holds. This is crucial because Python needs to know what operations it can perform on that data (e.g., you can add two numbers, but you cannot add a number to a name).

### Common Built-in Data Types:

| Category | Type | Description | Example |
| --- | --- | --- | --- |
| **Numeric** | `int`, `float`, `complex` | Whole numbers, decimals, or complex math numbers. | `5`, `3.14`, `8 + 2j` |
| **Text** | `str` (String) | A sequence of characters wrapped in `" "` or `' '`. | `"Harry"` |
| **Boolean** | `bool` | Represents logical values: Yes or No. | `True`, `False` |
| **None** | `NoneType` | Represents the absence of a value. | `None` |

---

## 3. Sequence & Mapped Data

Sometimes we need to store a collection of items rather than just one.

### List vs. Tuple

* **List:** A collection of items that is **Mutable** (can be changed).
* **Tuple:** A collection of items that is **Immutable** (cannot be changed).
> **Memory Trick:** "Mutation" means change (like evolution). If it's *mutable*, you can mutate/change it!



### Dictionary (`dict`)

A collection of **key-value pairs**, similar to a real-world dictionary or a student report card.

* **Example:** `{"name": "Sakshi", "age": 20}`. Here, "name" is the key, and "Sakshi" is the value.

---

## 4. Key Functions & Concepts

* **`type()` function:** Use this to find out the data type of any variable.
* *Example:* `print(type(a))` might output `<class 'int'>`.


* **Everything is an Object:** In Python, every piece of data—whether a number or a list—is technically an "Object" belonging to a "Class."

---

### Important Rule: Same-Type Operations

You generally cannot perform operations between incompatible types.

* `print(5 + 5)` ✅ (Result: 10)
* `print("5" + "5")` ✅ (Result: "55" - this is called **concatenation**)
* `prinamples of how to convert one data type into another (Typecasting)?**

lecture 7

In [56]:
print (5+6)
print(2*9)
print(81/9)
print(81//8)
print(12-8)
print(8%2)
print(8%3)
print(5**3)

11
18
9.0
10
4
0
2
125


In [57]:
print(81//8)

10


# Summary Notes: Python Operators & Calculator Challenge

This session focuses on basic arithmetic operators and sets up your first practical coding challenge.

---

## 1. Arithmetic Operators in Python

Python supports all standard mathematical operations. Here is a quick reference:

| Operator | Name | Description | Example | Result |
| --- | --- | --- | --- | --- |
| `+` | **Addition** | Adds two numbers | `15 + 7` | `22` |
| `-` | **Subtraction** | Subtracts one from another | `15 - 7` | `8` |
| `*` | **Multiplication** | Multiplies two numbers | `15 * 6` | `90` |
| `/` | **Division** | Divides and gives a float | `15 / 6` | `2.5` |
| `//` | **Floor Division** | Divides and removes the decimal | `15 // 6` | `2` |
| `%` | **Modulo** | Gives the **remainder** of a division | `5 % 3` | `2` |
| `**` | **Exponential** | Power operator () | `5 ** 3` | `125` |

---

## 2. Key Concept: Floor Division vs. Modulo

* **Floor Division (`//`)**: Useful when you only want the whole number part of a division. It "chops off" the decimal.
* **Modulo (`%`)**: Extremely useful for checking if a number is even or odd (e.g., `x % 2 == 0`) or finding remainders.

---

## 3. Exercise: Create a Calculator

**Objective:** Build a simple Python script that performs basic arithmetic on two numbers.

**Requirements:**

1. Create two variables (e.g., `a = 50` and `b = 3`).
2. Print the results of **Addition, Subtraction, Multiplication, and Division** using these variables.
3. (Bonus) Include **Floor Division, Modulo, and Exponential** results.

**How to Submit:**

* Create a new Repl on [Replit](https://replit.com).
* Write your code and ensure it runs using the "Run" button.
* Share the link in the comments of the lesson to get a potential shout-out!

---

### Pro Tip

Remember, the goal isn't just to be the fastest; it's to understand how these operators work. The real competition is **"You vs. You"**—just focus on improving your skills one day at a time.


Lecture 8

Lecture 9

In [58]:
print("shashi"+9)

TypeError: can only concatenate str (not "int") to str

In [59]:
print("shashi"+"kumar")

shashikumar


In [60]:
print(8+9)


17


In [61]:
print(True+True)


2


In [62]:
print(True+False)

1


In [64]:
print(False+True)

1


In [63]:
print(False+False)

0


In [65]:
a = "shashi"
b="kumar"
print(a+b)

shashikumar


In [66]:
a = 2
b=4
print(a+b)

6


In [67]:
a = "2"
b="4"
print(a+b)

24


In [68]:
print (int(a) + int (b))

6


In [70]:
a = "2shashi"
b="4"
print (int(a) + int (b))

ValueError: invalid literal for int() with base 10: '2shashi'

In [72]:
a = "2shashi"
b="4"
print (str(a) + str (b))

2shashi4


In [73]:
c =1.9
d=9
print(c+d)

10.9


# Summary Notes: Type Casting in Python

Day #9 is all about **Type Casting** (also known as Type Conversion). This is the process of converting one data type into another to perform specific operations or avoid errors.

---

## 1. What is Type Casting?

Type casting allows you to change a variable from one type (e.g., a **String**) to another (e.g., an **Integer**).

**The Problem:**
If you have `a = "1"` and `b = "2"`, Python treats them as text.

* `print(a + b)` results in **"12"** (Concatenation).

**The Solution:**

* `print(int(a) + int(b))` results in **3**.

---

## 2. Explicit Type Casting

This is done **manually** by the programmer using built-in functions. You are telling Python exactly what to do.

* **Common Functions:**
* `int()`: Converts to an Integer.
* `float()`: Converts to a Floating-point number.
* `str()`: Converts to a String.
* `hex()`, `oct()`: Converts to Hexadecimal or Octal.



> **Warning:** The content must be valid for the conversion. You can convert `"27"` to an integer, but you cannot convert `"Harry"` to an integer—this will cause a **ValueError**.

---

## 3. Implicit Type Casting

This is done **automatically** by the Python interpreter to prevent data loss. Python converts a "lower" data type (like an Integer) to a "higher" data type (like a Float) during calculations.

**Example:**

```python
c = 1.9  # Float
d = 8    # Integer
print(c + d) # Result: 9.9 (Float)

```

In this case, Python automatically promoted the integer `8` to the float `8.0` so it could add it to `1.9` without losing the decimal information.

---

## 4. Summary Table

| Feature | Explicit Type Casting | Implicit Type Casting |
| --- | --- | --- |
| **Who does it?** | The Programmer | The Python Interpreter |
| **Methods used** | `int()`, `float()`, `str()`, etc. | Automatic (no method needed) |
| **Purpose** | To force a specific data type | To avoid data loss during math |

---

### Key Takeaway

Always be mindful of your data types! If your math looks like `"1" + "2" = "12"`, you likely need to use **Explicit Type Casting** to turn those strings into numbers.



Lecture 10

In [74]:
a =input()
print(a)

 67


67


In [76]:
a=input("enter your name:")
print("my name is",a)

enter your name: shashi kumar


my name is shashi kumar


In [78]:
x=input("enter your fname:")
y=input("enter your lname:")
print(x+y)

enter your fname: shashi
enter your lname: suman


shashisuman


In [79]:
x=input("enter your fnumber:")
y=input("enter your lnumber:")
print(x+y)

enter your fnumber: 12
enter your lnumber: 50


1250


In [80]:
x=input("enter your fnumber:")
y=input("enter your lnumber:")
print(int(x)+int(y))

enter your fnumber: 50
enter your lnumber: 70


120


In [81]:
x=input("enter your fnumber:")
y=input("enter your lnumber:")
print(int(x)+int(y))

enter your fnumber: shashi
enter your lnumber: 45


ValueError: invalid literal for int() with base 10: 'shashi'

# Summary Notes: Taking User Input in Python

Today, we take our programs from static to interactive! Input is the reason games and apps are addictive—they respond to **you**.

---

## 1. The `input()` Function

In Python, we use the `input()` function to get data from the user.

* **How it works:** When Python reaches an `input()` line, it pauses the program and waits for the user to type something and press **Enter**.
* **Variable Storage:** You must assign the input to a variable to save what the user typed.

**Example:**

```python
name = input("Enter your name: ")
print("Hello", name)

```

---

## 2. The "Greatest Doubt": String vs. Number

This is where most beginners get stuck. **The `input()` function always returns a String.**

### The Problem:

If you ask for two numbers and try to add them, Python treats them like text.

* User enters: `12` and `100`
* `print(x + y)` → Result: `"12100"`
* This is called **String Concatenation** (sticking text together side-by-side).

### The Solution:

If you want to do math, you must **Type Caste** the input into an integer or float immediately.

**Correct Way:**

```python
x = input("Enter first number: ")
y = input("Enter second number: ")
print(int(x) + int(y)) # Converts strings to integers before adding

```

---

## 3. Handling Errors (The "Value Error")

Python isn't a magician. If you tell Python to convert a word like `"Harry"` into an integer (`int("Harry")`), the program will crash.

* **Error Message:** `ValueError: invalid literal for int()`
* **Meaning:** Python is saying, "I can convert '12' to a number, but I have no idea how to turn 'Harry' into a number!"

---

## 4. Exercise: Interactive Calculator

It's time to upgrade the calculator you built in Exercise 1!

**Your Task:**

1. Create a program that asks the user for **two numbers**.
2. Perform **Addition, Subtraction, Multiplication, and Division**.
3. Print the results clearly (e.g., `"The sum of your numbers is: 112"`).

**Goal:** Practice combining `input()` with `int()` or `float()` type casting.

---

### Key Takeaway

`input()` = Text (always). If you need numbers, wrap your variables in `int()` or `float()`!


